src/v2/content_digest.rs

```diff
@@ -1,36 +1,47 @@
-use crate::errors::Result;
 /// Implements types and methods for content verification
 use sha2::{self, Digest};
 
 /// ContentDigest stores a digest and its DigestAlgorithm
 #[derive(Clone, Debug, PartialEq)]
-pub(crate) struct ContentDigest {
+pub struct ContentDigest {
     digest: String,
     algorithm: DigestAlgorithm,
 }
 
 /// DigestAlgorithm declares the supported algorithms
 #[derive(Display, Clone, Debug, PartialEq, EnumString)]
-enum DigestAlgorithm {
+pub enum DigestAlgorithm {
     #[strum(to_string = "sha256")]
     Sha256,
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum ContentDigestError {
+    #[error("digest {0} does not have algorithm prefix")]
+    BadDigest(String),
+    #[error("unknown algorithm")]
+    AlgorithmUnknown(#[from] <DigestAlgorithm as std::str::FromStr>::Err),
+    #[error("verification failed: expected '{expected}', got '{got}'")]
+    Verify {
+        expected: ContentDigest,
+        got: ContentDigest,
+    },
+}
+
 impl ContentDigest {
     /// try_new attempts to parse the digest string and create a ContentDigest instance from it
     ///
     /// Success depends on
     /// - the string having a "algorithm:" prefix
     /// - the algorithm being supported by DigestAlgorithm
-    pub fn try_new(digest: String) -> Result<Self> {
+    pub fn try_new(digest: String) -> std::result::Result<Self, ContentDigestError> {
```

Comment:
What's the advantage of declaring this specific error type in this module vs. adding the variants to the crate's error module?

Reply:
If one wishes to use specifically ContestDigest, they can exhaustively match on ContestDigestError and handle all possible errors. If this function returns `crate::Error` instead, such a user would have to 
1) add wildcard check
2) write justyfing comment
3) carefully read through docs of `dkregistry::Error` (or source code) and decide for each variant if it is possible.
It is time-consuming process.

In short: IMHO, if some function returns E as error enum, and it contains variants that are never returned by this function, it is inconvenient for API user who would like to do exhaustive match.


That said, I'm ready to merge this domain-specific small enums into `crate::Error`, if you find that behavior more maintainable or readable,

Reply:
Thanks for elaborating, the reasoning makes sense to me! It would be nice if Rust allowed a convenient way to specify a subset of variants the of the error enum a function returns, but I can't come up with an easy way to do that. In this case I prefer the specific error type as implemented in this PR.